sktime/utils/deep_equals/_deep_equals.py

```diff
@@ -103,7 +103,10 @@ def _fh_equals_plugin(x, y, return_msg=False, deep_equals=None):
         return ret(False, ".is_relative")
 
     # recurse through values of x, y
-    is_equal, msg = deep_equals(x._values, y._values, return_msg=True)
+    try:
```

Comment:
Generally, I think one should avoid try/except architecture like this unless one is explicitly concerned with error handling, as failures can have multiple reasons.

What I would do, is check `len(x._values)` against `len(x._values)` and add an early return.

Reply:
Generally, I would do the same. 
Here, I chose the try-catch as I have basically no guarantees on the type of `*._values`. If I am right the types of ForecastingHorizon._values are (int, list, np.ndarray, pd.Index). 

The issue is that  len(int(1))  results in an type error. As this is handled recursively and the error might pop up in pd.Index and np.ndarray and both result in an ValueError and are handled downstream (in skbase i guess), with regards to sktime the cleanest way is to return equality=False if an ValueError happens downstream.

Reply:
hmmm, I see.

If I understand the error correctly, `scikit-base` does not anticipate that `x==y` can result in an error, it only covers cases of this resulting in a `bool` or an iterable of `bool`.

How about we add another case, directly above `if isinstance(x == y, bool):`, that does `hasattr(x, `__len__`)` etc and then `len(x) == len(y)`? Only in `scikit-base`.

Reply:
just as sa thought.. I guess the correct way would be to handle it in the corresponding skbase _pandas_equals or _numpy_equals methods. Otherwise, we introduce even more duplication of logic and code in the end.

Reply:
FYI, `numpy` seems fine, since `np.array_equals` does the job of comparing shape first, values second.

Reply:
> just as sa thought..

Yes, we crossed channel here, we had the same idea. `numpy` seems fine, only `pandas` needs fixing.

I would still also add the catch for the `len` case too, to cover any other objects that might fail on `==` if lengths are different. Two separate PR?

Reply:
First part probably works. 
For the second I kind of doubt that the sktime plugins are ever run if the error occurs  (unless the plugin of skbase returns true we have an early return with false; But that's rather theoretical! ). For consistency it should the issue in the `_pandas_equals_plugin` should be solved in both cases.  How do you handle such cases (also with regards to deployment and versioning)?

Reply:
> For the second I kind of doubt that the sktime plugins are ever run

The plugins are run *before* the "catch-all logic" which has things like `ifinstance(x==y, bool)`, only stuff not caught by the plugins and base comparisons ends up here.

So we should be fine?

> How do you handle such cases (also with regards to deployment and versioning)?

Generally, bugfixes that do not break anything can go in PATCH versions of `scikit-base`, changes like the plugin refactor risk breaking dependent functionality and hence are released only with MINOR versions and have to go through deprecatoin/change management first.

* semantic versioning, MINOR, PATCH, see https://semver.org/
* deprecation/change policy: https://www.sktime.net/en/latest/developer_guide/deprecation.html

In this specific case, we'll have to raise lower bounds on `scikit-base` to a MINOR version after a while.

For `deep_equals`, we can be more aggressive though - it is used mostly in testing, and otherwise only in:

* the `BaseObject` equality dunder, probably we should de-duplicate with `scikit-base`
* one location in the `dists_kernel` module where functionality should not be impacted